Assignment 2
Author: Grismely M. Hiraldo


In [25]:
import numpy as np
import matplotlib.pyplot as plt 
from mpl_toolkits import mplot3d
from sklearn import datasets
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier

In [26]:
# import the Iris data set

iris = datasets.load_iris()
X = iris.data
Y = iris.target
Xzeros0 = (0. * X) + .2
Xzeros1 = (0. * X) 
Xzeros2 = (0. * X) - .2
x_min, x_max = X.min() - 0.5, X.max() + 1.5

# Classifications

In [28]:
# get Iris classifications

IC = np.unique(Y)
IrisC = ('Setosa','Versicolor','Virginica')
pltC = ('red', 'blue', 'green')
pltCback = ('pink', 'lightblue','lightgreen')
pltM = ('o', 'x', '*')
for i in range(0,3):
    print (IC[i],pltM[i],IrisC[i],pltC[i],pltCback[i])

0 o Setosa red pink
1 x Versicolor blue lightblue
2 * Virginica green lightgreen


# Features

In [29]:
# setup Iris feature lablels

IrisF=("sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)")
for i in range(0,4):
    print(IrisF[i])


sepal length (cm)
sepal width (cm)
petal length (cm)
petal width (cm)


In [30]:
# now choose the classes and features to be modeled.

# note that if more than two classes are specified, scikit learn will use a multiclass model.

# Two Classes 

## A) Setosa and Versicolor

### One Feature

#### 1. Sepal Length (cm)

In [31]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.5


The accuracy of the model on the subset of the data used for this pair is only .5, which can be intereppted as the model not being very accurate on this particular combination of features.

#### 2. Sepal Width (cm)

In [32]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))



Score:  0.82


The accuracy of the model on the subset of the data used for this pair is 0.82, this is signficantly higher than it was for sepal length as the feature, the model was able to more accurately recognize this feature combination than the previous one.

#### 3. Petal Length (cm)

In [33]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The first 1.0 score, up until this point this would be the highest score and most accurate occurance of the model over a subdata combination. 

#### 4. Petal  Width (cm)

In [34]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The accuracy of the model on the subset of the data used for this pair is 1.0, this score of 100% tells us that that the overall model could be a good model despite the lower score with the first pair as we are now starting to see 100% acuracy over some of the subsets/feature combinations.

### Two Features

#### 5. Sepal Length & Sepal Width

In [35]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.99


The accuracy of the model on the subset of the data used for this pair is 0.99 it is nearly 100% accurate which furter supports my claim that the model may be an overall good model as we were seeing 100% accuracy or at high scores. This is the first of the two feature subset so it also may result that the model is trained more accurately over the two feature subset than the single feature subset.

#### 6. Sepal Length & Petal Length

In [36]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


A score of 1.0 meaning 100% accuracy for the model over this subset of the data, I think at this point in time I would assume that the model will reveal to have 100% or nearing 100% accuracy for the rest of the subset.

#### 7.Sepal Length & Petal Width

In [37]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The score supports my previous claim of the model over this particular subset.

#### 8. Sepal width & Petal length

In [38]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The score supports my previous claim of the model over this particular subset.

#### 9. Sepal Width & Petal Width

In [39]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The score supports my previous claim of the model over this particular subset.

#### 10. Petal Length & Petal Width

In [40]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The final pair in the subset has a score of 1.0 accuracy. Therefore it can be said for this subset the model was accurate and I also suspect that there may be some underfiting in regards two particulat features sepal width and sepal length, I think its important to note that the least accurate scores were from subsets that included these features. In the single feature subset sepal length and width had less accurate scores comapred to the other features and in the two feature subset the only score that wasnt 1.0 for the model was the pair that included BOTH the sepal width and length. 

### Three Features

#### 11. Sepal Length, Sepal Width & Petal Length

In [41]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The accuracy of the model on the subset of the data is 1.0, this is the first pair in the three feature subset but this score has my second hypothesis under review about Sepal length and width being underfitted, they were both features in this pair and the model was 100% accurate so my hypothesis may only be true for single and double feature subset. This score does however support my overall hypothesis that the model was trained well over the data as I continue to see high reports of accuracy of the model over all subsets.

#### 12. Sepal Length, Sepal Width & Petal Width

In [42]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 13. Sepal Length, Petal Length & Petal Width

In [43]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was also 100% accurate over this subset pair.

#### 14. Sepal Width, Petal Length & Petal Width

In [44]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

### Four Features

#### 15. Sepal Length, Sepal Width, Petal Length, & Petal Width

In [45]:
# choose the classes and features to be included in the model

cdx = np.array([0,1])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1

# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0) # creating the model

#XX is the features that we are training on the classifications of all the 150 samples. YY contains all the smaples
#num of XX = num of YY
clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0



SL = Sepal Length <br>
SW = Sepal Width <br>
PL = Petal Length <br>
PW = Petal Width <br>
    
    
## Result

| Features | Accuracy |
| --- | --- |
| SL      | 0.5 |
| SW      | 0.82 |
| PL      | 1.0 |
| PW      | 1.0 |
| SL - SW | 0.99 |
| SL - PL | 1.0 |
| SL - PW | 1.0 |
| SW - PL | 1.0 |
| SW - PW | 1.0 |
| PL - PW | 1.0 |
| SL - SW - PL | 1.0 |
| SL - SW - PW | 1.0 |
| SL - PL - PW | 1.0 |
| SW - PL - PW | 1.0 |
| SL - SW - PL - PW | 1.0 |

The results shows that there is overall accuracy of the model over this subset with the excetion of "0.5" which serves as an outliter. The more features were included the higher the accuracy score it seems.

## B) Setosa and Virginica

### One Feature

#### 1. Sepal Length (cm)

In [46]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.72


The score of 0.72 is again one of the lower accuracy scores of the model over any of the subsets. Now its starting to look like a pattern single features seem to have less accuracy but specifically sepal legnth.

#### 2. Sepal Width (cm)

In [47]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.52


Again we see the model to be less accurate over sepal length and width and again over a single feature, this alludes to potential flaws in the model's training.

#### 3. Petal Length(cm)

In [48]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 4. Petal Width(cm)

In [49]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

### Two Features

#### 5. Sepal Length & Sepal Width

In [50]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.99


A score of 0.99 says the model was nearly compeltely accurate over this subset pair, this pair consist of both sepal length and width but the distinguishing feature that probably made it more accurate is that while the model was not as accurate over two features INDIVIDUALLY as a PAIR it is, an andiciation that more features more accuracy may be true for our model.


#### 6. Sepal Length & Petal Length

In [51]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 7.Sepal Length & Petal Width

In [52]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 8. Sepal width & Petal length

In [53]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair, and again appears to be genrally a good model with some slight discrepencies.

#### 9. Sepal Width & Petal Width

In [54]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 10. Petal Length & Petal Width

In [55]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

### Three Features

#### 11. Sepal Length, Sepal Width & Petal Length

In [56]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 12. Sepal Length, Sepal Width & Petal Width

In [57]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 13. Sepal Length, Petal Length & Petal Width

In [58]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

#### 14. Sepal Width, Petal Length & Petal Width

In [59]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0


The model was 100% accurate over this subset pair.

### Four Features

#### 15. Sepal Length, Sepal Width, Petal Length, & Petal Width

In [60]:
# choose the classes and features to be included in the model

cdx = np.array([0,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  1.0



SL = Sepal Length <br>
SW = Sepal Width <br>
PL = Petal Length <br>
PW = Petal Width <br>
    
    
## Result

| Features | Accuracy |
| --- | --- |
| SL      | 0.72 |
| SW      | 0.52 |
| PL      | 1.0 |
| PW      | 1.0 |
| SL - SW | 0.99 |
| SL - PL | 1.0 |
| SL - PW | 1.0 |
| SW - PL | 1.0 |
| SW - PW | 1.0 |
| PL - PW | 1.0 |
| SL - SW - PL | 1.0 |
| SL - SW - PW | 1.0 |
| SL - PL - PW | 1.0 |
| SW - PL - PW | 1.0 |
| SL - SW - PL - PW | 1.0 |

Our results continue to support my now (3) hypothesis of sorts. (1) That our model is generally a good model as we see overhwleming scores of 1.0, (2) there may be some kind of underfiting with the features of sepal length and sepal width as our least accurate scores always originate from the two, (3) there is a correlation between amount of features and accuracy, the two features with the least accurate results are always a pair of two classes and only one feature.

## C) Versicolor and Virginica

### One Feature

#### 1. Sepal Length (cm)

In [61]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.51


The accuracy of the model over this subset pair was only 0.51,the mdoel was not very accuracte over this pair.

#### 2. Sepal Width (cm)

In [62]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.6


The accuracy of the model over this subset pair was only 0.6,the mdoel was not very accuracte over this pair.

#### 3. Petal Length(cm)

In [63]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.84


The accuracy of the model over this subset pair was, surpisngly, only 0.84, in the previous data sets we see more accuracy for petal length and width. However, this may be a result of the classes, something that I hadn't fully considered until this subset.

#### 4. Petal Width(cm)

In [64]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.84


The score of 0.84 is a relatively high accuracy score but in comparassion to the previous scores, the classes seems to have affected the accuracy significantly.

### Two Features

#### 5. Sepal Length & Sepal Width

In [65]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.74


This score of only 0.74 dampers my third hypothesis, there was an increase in features but that did not translate to an increate in accuracy for this pair.

#### 6. Sepal Length & Petal Length

In [66]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.92


An increase in features DID translate to an increate in accuracy for this pair.

#### 7.Sepal Length & Petal Width

In [67]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.63


But not for this pair.

#### 8. Sepal width & Petal length

In [68]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.91


#### 9. Sepal Width & Petal Width

In [69]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.87


#### 10. Petal Length & Petal Width

In [70]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.9


### Three Features

#### 11. Sepal Length, Sepal Width & Petal Length

In [71]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.89


#### 12. Sepal Length, Sepal Width & Petal Width

In [72]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.91


#### 13. Sepal Length, Petal Length & Petal Width

In [73]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.75


#### 14. Sepal Width, Petal Length & Petal Width

In [74]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.9


### Four Features

#### 15. Sepal Length, Sepal Width, Petal Length, & Petal Width

In [75]:
# choose the classes and features to be included in the model

cdx = np.array([1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.83



SL = Sepal Length <br>
SW = Sepal Width <br>
PL = Petal Length <br>
PW = Petal Width <br>
    
    
## Result

| Features | Accuracy |
| --- | --- |
| SL      | 0.51 |
| SW      | 0.6 |
| PL      | 0.84 |
| PW      | 0.84 |
| SL - SW | 0.74 |
| SL - PL | 0.92 |
| SL - PW | 0.63 |
| SW - PL | 0.91 |
| SW - PW | 0.87 |
| PL - PW | 0.9 |
| SL - SW - PL | 0.89 |
| SL - SW - PW | 0.91 |
| SL - PL - PW |0.75  |
| SW - PL - PW | 0.9 |
| SL - SW - PL - PW | 0.83 |

The results of the model over this data subset show less accuracy over the combination of these two classes. I think features may still play a role but now I must consider the role the classes play in accuracy as this combination of classes did not result in the model having 100% accuracy. Instead there was alot of variety in the range of scores.

# Three Classes 

## Setosa, Versicolor and Virginica

### One Feature

#### 1. Sepal Length (cm)

In [76]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.68


The pairing of three classes and a single feature resulted in the model not being as accuracte over the subset pair, this is not surprising as this is a consisent pattern seen over the data especially in regards to this specific feature.

#### 2. Sepal Width (cm)

In [77]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.52


The pairing of three classes and a single feature resulted in the model not being as accuracte over the subset pair, this is not surprising as this is a consisent pattern seen over the data especially in regards to this specific feature.

#### 3. Petal Length(cm)

In [78]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.9466666666666667


The model was accurate over this pairing potentially due to the specific feature.

#### 4. Petal Width(cm)

In [79]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.96


The model was accurate over this pairing potentially due to the specific feature.

### Two Features

#### 5. Sepal Length & Sepal Width

In [80]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.7733333333333333


Three feature pairings should more accurasy in the previous data subset and this seems to hold true for this one, although not 100% accurate the score is higher than the previous subset pairs.

#### 6. Sepal Length & Petal Length

In [81]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.8733333333333333


As stated previously pairs with more features seem to show relatively more accuracy  but the inclusion of more classes actually changes my mind about the idea that our model is a generally good model, it appears to be a simple model, as it did best with pairs of two classes and specific class combinations.

#### 7.Sepal Length & Petal Width

In [82]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.94


#### 8. Sepal width & Petal length

In [83]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.8866666666666667


#### 9. Sepal Width & Petal Width

In [84]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.8933333333333333


#### 10. Petal Length & Petal Width

In [85]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.96


### Three Features

#### 11. Sepal Length, Sepal Width & Petal Length

In [86]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.8866666666666667


#### 12. Sepal Length, Sepal Width & Petal Width

In [87]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.9266666666666666


#### 13. Sepal Length, Petal Length & Petal Width

In [88]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.7933333333333333


#### 14. Sepal Width, Petal Length & Petal Width

In [89]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.9133333333333333


### Four Features

#### 15. Sepal Length, Sepal Width, Petal Length, & Petal Width

In [90]:
# choose the classes and features to be included in the model

cdx = np.array([0,1,2])             # this array contains the classes to be considered (can be 2-3 classes)
fdx = np.array([0,1,2,3])           # this array contains the features to be considered (can be 1-4 features)

nC = len(cdx)
nF = len(fdx)

# build the XX and YY arrays
# these arrays contain the samples in the classes specified in cdx and include the features specified in fdx

i = 0
numY = 0
for i in range(len(Y)):
    for j in range(nC):
        if Y[i] == IC[cdx[j]]:
            numY += 1

YY = np.zeros((numY))
XX = np.zeros((numY,nF))
if (len(fdx) == 1):
    XXzeros0 = (0. * XX) + .2
    XXzeros1 = (0. * XX) 
    XXzeros2 = (0. * XX) - .2

j = 0
for i in range(len(Y)):
    for k in range(nC):
        if Y[i] == IC[cdx[k]]:        
            YY[j] = IC[cdx[k]]
            for m in range(nF):
                XX[j,m] = X[i,fdx[m]] 
            j += 1
            
# fit the model

# clf = Perceptron(tol=1e-4, random_state=0)
clf = SGDClassifier(tol=1e-4, random_state=0)

clf.fit(XX,YY)

print('\nScore: ',clf.score(XX,YY))


Score:  0.9666666666666667


# SL = Sepal Length <br>
SW = Sepal Width <br>
PL = Petal Length <br>
PW = Petal Width <br>
    
    
## Result

| Features | Accuracy |
| --- | --- |
| SL      | 0.68 |
| SW      | 0.52 |
| PL      | 0.9466666666666667 |
| PW      | 0.96 |
| SL - SW | 0.7733333333333333 |
| SL - PL | 0.8733333333333333 |
| SL - PW | 0.94 |
| SW - PL | 0.8866666666666667 |
| SW - PW | 0.8933333333333333 |
| PL - PW | 0.96 |
| SL - SW - PL | 0.8866666666666667 |
| SL - SW - PW | 0.9266666666666666 |
| SL - PL - PW | 0.7933333333333333 |
| SW - PL - PW | 0.9133333333333333 |
| SL - SW - PL - PW | 0.9666666666666667 |

The result of the model over this data subset shows that for this particular subset the model shows more accuracy with more features but alludes to bigger issues with the model.

## Discussion

The varying of the combinations of features used and the data included affect the accuracy of the model. Orginally I believed this to be exclusive to features but the combination of the features and the classes together ultiamtely affected the model. This is notable as the model was not as accurate for the subset's of βγ and αβγ. The scores were signficantly lother than the other data, the model was not trained as well over these two. Additionally, depending on the feature, specifically sepal length and width you would see signficantly lower accuracy scores when they were a single feature. To reflect on my previous assumptions: (1) That our model is generally a good model as we see overhwleming scores of 1.0, I think it is too naive to describe our model as a "generally good" model when we see such variance depending on the classes and features, however, the model did show accuracy over some data, so it is not a total failure. (2) there may be some kind of underfiting with the features of sepal length and sepal width as our least accurate scores always originate from the two. I still believe this to be true because it was a cosnsitent pattern throughout our data and subsets. (3) There is a correlation between amount of features and accuracy, the two features with the least accurate results are always a pair of two classes and only one feature. I believe this to depend on the data (class combinations).